# TP 2 - Familiarisation avec Python et Jupyter Notebook

Dans ce TP, vous allez écrire vos premières lignes de code **en Python**.

Pour cela, vous utilisez **Jupyter Notebook** : un carnet de notes qui vous permet d'exécuter du code Python depuis une page web sans avoir à installer quoique ce soit sur votre machine.

Jupyter Notebook est uniquement un environnement de travail à **but pédagogique**. Il est utilisé dans le monde de l'éducation pour enseigner le Python, l'algorithmique, la Data Science, etc. Il est également utilisé dans le **monde de la recherche** pour présenter des travaux et tester différentes approches.

En revanche, Jupyter Notebook n'est *pas* utilisé dans le monde de l'édition logicielle ou dans l'industrie. A la place, les développeurs installent sur leur machine un **environnement de développement** pour travailler. Dans un prochain TP, nous explorerons également cette approche.

Il existe 2 types de zones dans Jupyter :
* les **zones de texte** (Markdown): la zone actuelle *est* une zone de texte,
* les **zones de programmation** (Code) : identifiable aux encadrés gris et au préfix `In [ ]`.

Il est possible d'activer une zone en cliquant dessus (clic gauche simple).

Cliquez sur la zone de programmation ci-dessous (où il est écrit `print("Boujour, monde")`). Observez l'activation de cette zone, qui est reflétée par un encadré supplémentaire. Cette zone de programmation contient déjà un programme Python très simple. Pour l'exécuter, vous devez cliquer sur le bouton :arrow_forward: **Run** situé dans le menu en haut.

In [56]:
print("Bonjour, monde")

Bonjour, monde


L'exécution du code ci-dessus affiche "Bonjour, monde". Un programme "Hello, World" en anglais est souvent le premier programme que l'on écrit pour découvrir un nouveau langage de programmation.

Cela permet d'afficher dans la sortie standard de la console le texte écrit entre guillemets. Nous reviendrons sur les notions de sortie standard et de console dans le prochain cours. A ce stade, il est simplement important de comprendre et de retenir que vous venez d'exécuter votre premier programme Python :+1:.

Maintenant, cliquez dans l'encadré ci-dessous et cliquez sur **Run**.

In [57]:
1

1

Vous voyez que 1 est affiché en sortie. Jupyter renvoie le résultat de l'exécution de la dernière instruction qu'il trouve.

Pour vous en convaincre, exécutez le programme ci-dessous :

In [58]:
1
2

2

Cette fois-ci, la sortie est `2`, car il s'agit de la dernière instruction.

## Les opérateurs arithmétiques

Nous allons maintenant faire un peu de calcul basique :

In [59]:
1 + 1

2

Sans surprise, le résultat est `2`.

Evidemment, on peut aussi soustraire 2 nombres avec l'opérateur `-` :

In [60]:
6 - 4

2

Ce qui nous donne `2` à nouveau.

Les nombres négatifs sont précédés du caractère `-` :

In [61]:
-4 - 8

-12

On obtient `-12`.

Le caractère `*` représente une multiplication dans la plupart des langages de programmation :

In [62]:
2 * 3

6

Encore une fois, pas de grande surprise : on s'attend à voir `6`.

Le caractère `\` représente une division :

In [63]:
3 / 2

1.5

Cette fois-ci, vous devriez obtenir `1.5`. Notez que le caractère `.` est utilisé pour séparer la partie entière de la partie décimale. Cette convention vient de la langue anglo-saxonne qui utilise cette représentation des nombres réels. Cette différence est importante avec les mathématiques telles qu'elles sont enseignées en France et qui utilise le caractère `,` à la place.

En résumé, là où un français dira "un virgule cinq", un anglo-saxon dira "one dot five", c'est-à-dire "un point cinq". La quasi-totalité des langages de programmation utilisent la notation anglo-saxonne avec un point.

Tant que l'on parle de division, on vous a toujours expliqué que la division par zéro était impossible. Soyons perfides :imp: 1 minute et regardons si on peut faire une blague à Python :

In [64]:
1 / 0

ZeroDivisionError: division by zero

Des messages d'erreur s'affichent. Ne vous inquiétez pas, vous n'avez rien cassé ! Python a détecté que vous faisiez une opération de division par zéro et vous en informe avec une exception `ZeroDivisionError`.

Nous reviendrons sur les exceptions dans un prochain cours.

Légérement plus compliqué, le symbole `\\` permet d'effectuer une division entière. Cela signifie que la valeur renvoyée est la partie entière du résultat de la division :

In [ ]:
3 // 2

Cela devrait vous renvoyer `1`.

Prenons un autre exemple :

In [ ]:
23 // 3

Cela devrait vous renvoyer `7` car :

In [ ]:
23 / 3

Un autre concept important est le modulo, représenté par le symbole `%`. Il renvoit le reste de la division :

In [ ]:
23 % 3

Expliquez la relation entre ces 3 instructions :

In [ ]:
23 // 3

In [ ]:
23 % 3

In [ ]:
3 * 7 + 2

Il est d'ailleurs bien évidemment possible de combiner des opérations, comme on peut le voir dans le dernier exemple ci-dessus, et également dans l'exemple suivant :

In [ ]:
2 * 3 + 1

Les opérations sont priorisés généralement comme en mathématiques. Par exemple, la multiplication a la priorité sur l'addition :

In [ ]:
1 + 2 * 3

7

Il est possible d'utiliser des paranthèses pour changer l'ordre d'exécution, comme en mathématiques :

In [ ]:
(1 + 2) * 3

9

In [ ]:
3 * (23 // 3) + (23 % 3)

23

In [ ]:
3 * 23 // 3 + 23 % 3

L'opérateur `**` permet d'élever un nombre à une puissance. Donc $2^3$ devient :

In [ ]:
2 ** 3

8

Jusqu'à présent, nous avons manipulé uniquement des nombres entiers naturels IN et des nombres réels IR. En informatique, on parle de **types** :
* `integer` (entier) pour la représentation des nombres entiers naturels,
* `floating-point` (à point flottant) pour la représentation des nombres réels.

Sans rentrer dans les détails, il s'agit uniquement d'une sous-partie de ces ensembles. En effet, les limites physiques d'une machine contraignent l'espace que les valeurs peuvent prendre.

Voici une illustration. Quel devrait être les résultats de ces calculs ?

In [ ]:
1 / 3

0.3333333333333333

In [ ]:
100.01 + 100.01 + 100.01

300.03000000000003

Dans le premier cas, vous vous doutiez que le résultat allait forcément être une approximation puisque l'on sait que 1 / 3 fait 0.3333333333333333333333...

Dans le deuxième cas, il y a de quoi être vraiment surpris. Le nombre 300.03 ne parait pas compliqué, et pourtant ce n'est pas le résultat retourné. On obtient 300.03000000000003 à la place.

Dans les 2 cas, on appelle **approximation numérique** le résultat renvoyé. Dans le 2e cas, la valeur 0.00000000000003 est appelée **erreur d'approximation numérique**. Ces approximations sont une différence essentielle entre les mathématiques, qui ont une représentation "parfaite", et l'informatiqe qui a une représentation limitée par les capacités des machines et les normes de représentation des nombres.

Ces limitations et ces approximations ne devraient pas vous inquiéter : le champ d'action est vaste et permet de résoudre une très grande quantité de problèmes avec une très bonne précision.

## Opérateurs Booléen

Nous allons maintenant parler du type `bool` qui permet de représenter les types Booléen suivants :
* `True` pour vrai,
* `False` pour faux.

In [ ]:
True

True

In [ ]:
False

On retrouve les mêmes opérateurs de comparaison qu'en mathématiques.

Pour vérifier si un nombre est strictement plus petit qu'un autre :

In [ ]:
3 < 6

In [ ]:
6 < 3

Pour vérifier si un nombre est strictement plus grand qu'un autre :

In [ ]:
3 > 6

Pour vérifier si un nombre est plus petit ou égal, on utilise `<=` :

In [ ]:
2 <= 6

In [ ]:
2 <= 2

In [ ]:
2 <= 1

De manière similaire, on utilise `>=` pour l'inverse :

In [ ]:
2 >= 1

True

Pour comparer si 2 nombres sont égaux, on utilise `==`. Faites bien attention au double `==`. Il y a en a 2. Comme les 2 yeux.

On insiste lourdement ici à dessein : tout le monde oublie au moins une fois le double `==`. Un unique `=` a une signification complètement différente (nous y reviendrons plus tard dans ce TP).

Exécutez les exemples suivants :

In [ ]:
1 == 1

True

In [ ]:
3 == 4

False

In [ ]:
300.03 == 100.01 + 100.01 + 100.01

False

In [ ]:
1 = 1

SyntaxError: can't assign to literal (<ipython-input-25-ecdac3dcabfd>, line 1)

Les 2 premiers exemples ci-dessus sont triviaux et renvoient respectivement `True` et `False`.

Dans le 3e exemple, on illustre à nouveau le problème d'approximation numérique. On a donc 300.03 qui, au sens de la représentation de Python, n'est pas égal à 100.01 + 100.01 + 100.01.

Dans le 4e exemple, on a utilisé un unique `=`, juste pour montrer ce qui se passe dans ce cas précis. En pratique, l'interprêteur Python nous renvoie une exception de type `SyntaxError`. Cela signifie que ce code n'est pas valide car il n'obéit pas à la grammaire du langage Python. En tout cas, encore une fois, utilisez bien **deux (2)** symboles `==` lorsque vous voulez effectuer des comparaisons.

Très souvent, on souhaite que plusieurs conditions soient remplies en même temps. Par exemple, je souhaite que ma prochaine voiture soit verte *OU* rouge, *ET* qu'elle ait un moteur. En logique Booléenne, on appelle *OU* et *ET* s'appellent des **opérateurs logiques**.

En Python, on utilise tout simplement la version anglaise de ces mots : *or* et *and*

In [ ]:
1 < 2 and 2 < 3

True

In [ ]:
1 < 2 and 3 < 2

False

In [ ]:
1 < 2 or 3 < 2

True

In [ ]:
2 < 1 or 3 < 2

False

La combinatoire entre opérateurs logiques peut se résumer très simplement dans les matrices suivantes :

Ou :

| or    | True | False |
|-------|------|-------|
| True  | True | True  |
| False | True | False |

Et :

| and   | True  | False |
|-------|-------|-------|
| True  | True  | False |
| False | False | False |

## Variables

La notion de **variable** en informatique est très différente de celle en mathématiques. En mathématiques, on recherche régulièrement à connaître la valeur inconnue de variables `x`, `y`, `z`, etc. dans un système d'équations :

$2 (x + y)^2 = 3 z + 2$

En informatique, dans la quasi-totalité des langages de programmation, la valeur d'une variable à un instant donné est **toujours** connue et dépend uniquement du chemin d'exécution emprunté.

Certaines langages de programmation, notamment certains langages fonctionnels, les méthodes formelles ou les langages spécialisés dans les mathématiques, peuvent d'éloigner de cette règle et se rapprocher des mathématiques. Mais ces langages sont hors-sujets dans le cadre de ce cours.

En Python, **tout est objet**. On lie une valeur à une variable de la manière suivante :

In [ ]:
x = 3

On peut ensuite lire la valeur de x :

In [ ]:
x

3

Le caractère `=` est utilisé par désigner **une affectation** ou **assignement**. La valeur `3` est assignée à la variable `x`. Donc la variable `x` a pour valeur `3`. Désormais, à chaque fois que l'on utilise `x`, c'est comme si on avait utilisé la valeur `3` à la place :

In [ ]:
x < 5

True

Il est possible de changer la valeur assignée à `x`. On dit que `x` est mutable.

In [ ]:
x = 2 + 3

Maintenant, la valeur de `x` est `5`. Par conséquent :

In [ ]:
x < 5

False

Le résultat de cette simple inéquation a changé, puisque la valeur de x a changé.

Un programme Python est exécuté du haut vers le bas. Les valeurs des variables évoluent au cours de l'exécution.

Notez que l'on ne peut utiliser qu'une seule variable à gauche d'un `=`. En mathématiques, on pourrait écrire :

$x + y = 3$

En Python, cela donne :

In [ ]:
x + y = 3

SyntaxError: can't assign to operator (<ipython-input-36-c0588d07a3bb>, line 1)

Une erreur de syntaxte. La grammaire du langage Python (et de la plupart des langages de programmation) n'autorisent qu'une seule variable à gauche d'un assignement.

De plus, en Python, on ne peut pas directement résoudre une équation de ce type :

$2 (x + y)^2 = 3 z + 2$

Pour résoudre automatiquement ce type d'équation, il serait nécessaire d'écrire un **algorithme**.

Par contre, connaissant une solution :

$$
\begin{bmatrix} 
x \\
y \\
z \\
\end{bmatrix}
=
\begin{bmatrix} 
1 \\
0 \\
0 \\
\end{bmatrix}
$$

On peut facilement écrire un mini programme qui vérifie cette solution :

In [ ]:
x = 1
y = 0
z = 0
2 * (x + y) == 3 * z + 2

True

Notez bien que la dernière expression emploi un double `==` car on effectue une comparaison.

En général, on évite les noms de variable trop courts ou peu explicites. Le premier caractères d'une variable doit être un caractère alphabétique ASCII ou un `_`. Les autres caractères composant le nom d'une variable doivent être alphanumériques ASCII ou `_`.

Autrement dit, les noms suivants sont corrects :

In [ ]:
maVariable1 = 0

In [ ]:
Ma2eVariable = 1

In [ ]:
ma_3e_var = 2

In [ ]:
_maVar4 = 3

In [ ]:
_ = 4

En revanche, les noms suivants sont syntaxiquement incorrects :

In [ ]:
3DCube = 3

SyntaxError: invalid syntax (<ipython-input-43-6f3f89b0246c>, line 1)

In [ ]:
être@hôpital = 42

SyntaxError: can't assign to operator (<ipython-input-48-455c9eafb3da>, line 1)

In [ ]:
super-man = 3

SyntaxError: can't assign to operator (<ipython-input-44-ea5a0d1d2f9d>, line 1)

En Python, par convention, on choisit en général la syntaxe `snake_case`, qui signifie que les différents mots composant un nom de variable sont séparés par un `_`.

On préférera donc :

In [ ]:
carton_de_saumur_champigny = 6

à la place de :

In [ ]:
CartonDeSaumurChampigny = 6

## Exercices

Vous n'avez pas le droit d'utiliser de calculatrice, smartphone, ou tableau. Vous devez résoudre ces questions uniquement avec du code en Python.

Lorsque vous aurez terminé, vous pouvez sauvegarder vos résultats en utilisant le menu File > Download as > Notebook (ipynb).

### Exercice 1

Utilisez Python pour déterminer lequel de ces 2 nombres est le plus grand :
* $2^{100} + (12345 \times 67890)^2$
* $100000^{10} \times 999999$

In [ ]:
# Insérez votre réponse ici...
# Pour cela, vous pouvez supprimer ces commentaires 
# et les remplacer pour votre propre code.

True

### Exercice 2

Quel est le reste de la division de 987654321 par 42 ?

In [ ]:
# Insérez votre réponse ici...

23515579

Combien de fois a-t-on 42 dans 987654321 ?

In [ ]:
# Insérez votre réponse ici...

### Exercice 3

On calcule l'Indice de Masse Corporelle en divisant la masse en kg par le carré de la taille en centimètres.

Cela donne la formule :

$$imc = \frac{masse}{taille^2}$$

Définissez les variables taille et masse initialement à 80 et 180.
Définissez la variable imc en utilisant la formule ci-dessus.
Affichez True si la valeur d'imc est supérieure ou égale à 25.

In [65]:
# Insérez votre réponse ici...